src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,566 @@
+                        data_buf: &mut Vec<u8>)
+                        -> Result<()> {
+        let code = data.get_type_code();
+        try!(self.write_u8(code));
+        match *data {
+            // If the data has length in (0, 4], it could be inline here.
+            // And padding 0x00 to 4 bytes if needed.
+            Json::Literal(ref v) => {
+                try!(self.write_u8(*v));
+                let left = LENGTH_U32 - LENGTH_LITERAL;
+                for _ in 0..left {
+                    try!(self.write_u8(JSON_LITERAL_NIL));
+                }
+            }
+            _ => {
+                try!(self.encode_u32_le(*offset));
+                let start_len = data_buf.len();
+                try!(data_buf.encode_json_body(data));
+                *offset += (data_buf.len() - start_len) as u32;
+            }
+        };
+        Ok(())
+    }
+}
+impl<T: Write> JsonEncoder for T {}
+
+#[allow(dead_code)]
+pub trait JsonDecoder: NumberDecoder {
+    fn decode_json(&mut self) -> Result<Json> {
+        let code = try!(self.read_u8());
+        self.decode_json_body(code)
+    }
+
+    fn decode_json_body(&mut self, code_type: u8) -> Result<Json> {
+        match code_type {
+            TYPE_CODE_OBJECT => self.decode_json_obj(),
+            TYPE_CODE_ARRAY => self.decode_json_array(),
+            TYPE_CODE_LITERAL => self.decode_json_literal(),
+            TYPE_CODE_I64 => self.decode_json_i64(),
+            TYPE_CODE_DOUBLE => self.decode_json_double(),
+            TYPE_CODE_STRING => self.decode_json_str(),
+            _ => Err(invalid_type!("unsupported type {:?}", code_type)),
+        }
+    }
+
+    fn decode_json_obj(&mut self) -> Result<Json> {
+        // count size key_entries value_entries keys values
+        let element_count = try!(self.decode_u32_le()) as usize;
+        let total_size = try!(self.decode_u32_le()) as usize;
+        let left_size = total_size - LENGTH_U32 - LENGTH_U32;
+        let mut data = vec![0;left_size];
+        try!(self.read_exact(&mut data));
+        let mut obj = BTreeMap::new();
+        if element_count == 0 {
+            return Ok(Json::Object(obj));
+        }
+        // key_entries
+        let key_entries_len = LENGTH_KEY_ENTRY * element_count;
+        let mut key_entries_data = &data[0..key_entries_len];
+
+        // value-entry ::= type(byte) offset-or-inlined-value(uint32)
+        let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+        let mut value_entries_data = &data[key_entries_len..(key_entries_len + value_entries_len)];
+        let mut key_offset = key_entries_len + value_entries_len;
+        for _ in 0..element_count {
+            let key_real_offset = try!(key_entries_data.decode_u32_le());
+            let key_len = try!(key_entries_data.decode_u16_le());
+            let key_data = &data[key_offset..(key_offset + key_len as usize)];
+            let key = String::from(str::from_utf8(key_data).unwrap());
```

Comment:
s/unwrap()/try!(from_utf8())/
So that the data corruption would not lead to panic.